File: AI/AiTools.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace SimpleWars {
    /// <summary>
    /// システムは使わない，思考ルーチンを作るのにあるとうれしいような関数たちです．
	/// 拡充予定です
    /// </summary>
    class AiTools {

        // 攻撃行動可能な自ユニット・敵ユニット・攻撃位置のリストを ActionのListとして取得する
        public static List<Action> getAllAttackActions(int teamcolor, Map map) {
            List<Action> atkActions = new List<Action>();

            // 行動可能な自ユニットリスト
            List<Unit> myMovableUnits = map.getUnitsList(teamcolor, false, true, false);

            foreach (Unit myUnit in myMovableUnits) {
                // movableUnitsの攻撃行動リスト
                List<Action> atkActionsOfOne = RangeController.getAttackActionList(myUnit, map);

                // 全体に含める
                atkActions.AddRange(atkActionsOfOne);
            }

            return atkActions;
        }

		/// <summary>
		/// 可能な行動全てを取得する
		/// </summary>
        public static List<Action> getAllMoveActions(int teamcolor, Map map) {
            List<Action> moveActions = new List<Action>();

            // 行動可能な自ユニットリスト
            List<Unit> myMovableUnits = map.getUnitsList(teamcolor, false, true, false);

            //行動可能なユニットの移動可能箇所をリストにまとめる
            foreach (Unit myUnit in myMovableUnits) {
                bool[,] reachable = RangeController.getReachableCellsMatrix(myUnit, map);
                for (int i = 0; i < reachable.GetLength(0); i++) {
                    for (int j = 0; j < reachable.GetLength(1); j++) {
                        if (reachable[i, j]) {
                            moveActions.Add(Action.createMoveOnlyAction(myUnit, i, j));
                        }
                    }
                }
            }

            return moveActions;
        }

        // operationUnitがtargetUnitに攻撃可能かどうか．ダメージ効果0の場合にはfalseが返る
        public static bool isEffective(Unit operationUnit, Unit targetUnit) {
            if (operationUnit.getSpec().getUnitAtkPower(targetUnit.getTypeOfUnit()) != 0) { return true; }
            return false;
        }

        /// <summary>
        /// 現在の状態における全合法手を生成する
        /// </summary>
        public static List<Action> getAllActions(int teamColor, Map map) {
            List<Action> allActions = new List<Action>();

            allActions.AddRange(getAllAttackActions(teamColor, map));
            allActions.AddRange(getAllMoveActions(teamColor, map));

            return allActions;
        }


		//動いていないユニットの数を返す
		public static int getNumOfDontMovedUnit(int teamColor, Map map) {
			return map.getUnitsList(teamColor, false, true, false).Count;
		}

		//指定した危険度以下の行動を返す
		//危険度マップに関してはcalcRiskMapを参照
		public static List<Action> getDeffensiveActions(int teamcolor, Map map, int[,] riskOfMap, int LowThreshold, int HighThreshold) {
			List<Action> moveActions = new List<Action>();

			// 行動可能な自ユニットリスト
			List<Unit> myMovableUnits = map.getUnitsList(teamcolor, false, true, false);

			//行動可能なユニットの移動可能箇所をリストにまとめる
			foreach (Unit myUnit in myMovableUnits) {
				bool[,] reachable = RangeController.getReachableCellsMatrix(myUnit, map);
				for (int i = 1; i < reachable.GetLength(0); i++) {
					for (int j = 1; j < reachable.GetLength(1); j++) {
						if (reachable[i, j] && (riskOfMap[i, j] <= HighThreshold && riskOfMap[i, j] >= LowThreshold)) {
							moveActions.Add(Action.createMoveOnlyAction(myUnit, i, j));
						}
					}
				}
			}

			return moveActions;
		}

		/// <summary>
		/// 敵ユニットからの利き（攻撃範囲）に入ってる場所に，危険度を書き込む
		/// 危険度は敵何ユニットから攻撃される可能性があるかを示している，高いほどヤバイ（最大で4方向+敵の自走砲からの攻撃数となる）
		/// ※注意：計算方法がユニットの行動順序を考慮しない為，若干間違ってるときがある．　
		/// </summary>
		/// <param name="teamColor"></param>
		/// <param name="map"></param>
		/// <returns></returns>
		public static int[,] calcRiskMap(int teamColor, Map map) {
			int[] dx = new int[4] { +1, 0, -1, 0 };
			int[] dy = new int[4] { 0, -1, 0, +1 };

			List<Unit> enUnits = map.getUnitsList(teamColor, false, false, true);

			bool[, , ,] ifMap = new bool[map.getXsize(), map.getYsize(), enUnits.Count(), 5];
			int[,] riskOfMap = new int[map.getXsize(), map.getYsize()];

			int i = 0;

			foreach (Unit u in enUnits) {
				//直接攻撃タイプのユニットの場合は，
				if (u.getSpec().isDirectAttackType() == true) {
					bool[,] reachAbleCells = RangeController.getReachableCellsMatrix(u, map);

					for (int x = 1; x < map.getXsize() - 1; x++) {
						for (int y = 1; y < map.getYsize() - 1; y++) {
							//対応するセルの上下左右の場所をチェックする．
							for (int r = 0; r < 4; r++) {
								//上↓左右どこかが移動範囲ならば，(x,y)は攻撃可能な場所
								if (reachAbleCells[x + dx[r], y + dy[r]] == true) {
									ifMap[x, y, i, r] = true;
								}
							}
						}
					}
					//間接攻撃ユニットの場合
				} else {
					bool[,] attackAbleCells = RangeController.getAttackableCellsMatrix(u, map);

					for (int x = 1; x < map.getXsize() - 1; x++) {
						for (int y = 1; y < map.getYsize() - 1; y++) {
							//(x,y)が攻撃可能な場所の場合
							if (attackAbleCells[x, y] == true) {
								ifMap[x, y, i, 4] = true;
							}
						}
					}
				}
				i++;
			}

			for (int x = 1; x < map.getXsize() - 1; x++) {
				for (int y = 1; y < map.getYsize() - 1; y++) {
					//初期化
					bool[] TorF_Unit = new bool[enUnits.Count];
					bool[] TorF_Direction = new bool[4];
					int RiskOfUnit = 0;
					int RiskOfLine = 0;
					int RiskOfIndirection = 0;

					//対応するセルを攻撃可能なユニット数をチェックする．
					for (int k = 0; k < enUnits.Count; k++) {
						TorF_Unit[k] = ifMap[x, y, k, 0] || ifMap[x, y, k, 1] || ifMap[x, y, k, 2] || ifMap[x, y, k, 3];
						if (TorF_Unit[k]) RiskOfUnit++;
						if (ifMap[x, y, k, 4]) RiskOfIndirection++;
					}

					//対応するセルを攻撃可能な方向をチェックする．
					for (int r = 0; r < 4; r++) {
						for (int k = 0; k < enUnits.Count; k++) {
							TorF_Direction[r] |= ifMap[x, y, k, r];
						}
						if (TorF_Direction[r]) RiskOfLine++;
					}

					if (RiskOfLine > RiskOfUnit) {
						riskOfMap[x, y] = RiskOfUnit + RiskOfIndirection;
					} else {
						riskOfMap[x, y] = RiskOfLine + RiskOfIndirection;
					}

				}
			}

			return riskOfMap;
		}


		/// <summary>
		/// 敵ユニットからの利き（攻撃範囲）に入ってる場所に，危険度配列に書き込む
		/// </summary>
		/// <param name="teamColor"></param>
		/// <param name="map"></param>
		/// <returns></returns>
		public static int[,] calcIFMap(int teamColor, Map map) {
			int[] dx = new int[4] { +1, 0, -1, 0 };
			int[] dy = new int[4] { 0, -1, 0, +1 };

			int[,] ifMap = new int[map.getXsize(), map.getYsize()];
			List<Unit> enUnits = map.getUnitsList(teamColor, false, false, true);

			foreach (Unit u in enUnits) {
				//直接攻撃タイプのユニットの場合は，
				if (u.getSpec().isDirectAttackType() == true) {
					bool[,] reachAbleCells = RangeController.getReachableCellsMatrix(u, map);

					for (int x = 1; x < map.getXsize() - 1; x++) {
						for (int y = 1; y < map.getYsize() - 1; y++) {
							//対応するセルの上下左右の場所をチェックする．
							for (int r = 0; r < 4; r++) {
								//上↓左右どこかが移動範囲ならば，(x,y)は攻撃可能な場所
								if (reachAbleCells[x + dx[r], y + dy[r]] == true) {
									ifMap[x, y] += 1;
								}
							}
						}
					}
					//間接攻撃ユニットの場合
				} else {
					bool[,] attackAbleCells = RangeController.getAttackableCellsMatrix(u, map);

					for (int x = 1; x < map.getXsize() - 1; x++) {
						for (int y = 1; y < map.getYsize() - 1; y++) {
							//(x,y)が攻撃可能な場所の場合
							if (attackAbleCells[x, y] == true) {
								ifMap[x, y] += 1;

							}
						}
					}
				}
			}
			return ifMap;
		}
    }
}

File: port/AiTools.py
from Action import Action
from RangeController import RangeController

class AiTools:
    @staticmethod
    def get_all_attack_actions(teamcolor, map):
        """
        Get a list of all possible attack actions as a list of Action objects.
        """
        atk_actions = []

        # List of movable own units
        my_movable_units = map.get_units_list(teamcolor, False, True, False)

        for my_unit in my_movable_units:
            # List of attack actions for movable units
            atk_actions_of_one = RangeController.get_attack_action_list(my_unit, map)

            # Add to the overall list
            atk_actions.extend(atk_actions_of_one)

        return atk_actions

    @staticmethod
    def get_all_move_actions(teamcolor, map):
        """
        Get all possible move actions.
        """
        move_actions = []

        # List of movable own units
        my_movable_units = map.get_units_list(teamcolor, False, True, False)

        # Gather the movable positions of movable units into a list
        for my_unit in my_movable_units:
            reachable = RangeController.get_reachable_cells_matrix(my_unit, map)
            for i in range(len(reachable)):
                for j in range(len(reachable[0])):
                    if reachable[i][j]:
                        move_actions.append(Action.create_move_only_action(my_unit, i, j))

        return move_actions

    @staticmethod
    def is_effective(operation_unit, target_unit):
        """
        Check if operationUnit can attack targetUnit. Returns False if the damage effect is 0.
        """
        if operation_unit.get_spec().get_unit_atk_power(target_unit.get_type_of_unit()) != 0:
            return True
        return False

    @staticmethod
    def get_all_actions(team_color, map):
        """
        Generate all legal moves in the current state.
        """
        all_actions = []

        all_actions.extend(AiTools.get_all_attack_actions(team_color, map))
        all_actions.extend(AiTools.get_all_move_actions(team_color, map))

        return all_actions

    @staticmethod
    def get_num_of_dont_moved_unit(team_color, map):
        """
        Return the number of units that have not moved.
        """
        return len(map.get_units_list(team_color, False, True, False))

    @staticmethod
    def get_deffensive_actions(teamcolor, map, risk_of_map, low_threshold, high_threshold):
        """
        Return actions with the specified risk level or below.
        For the risk map, refer to calc_risk_map.
        """
        move_actions = []

        # List of movable own units
        my_movable_units = map.get_units_list(teamcolor, False, True, False)

        # Gather the movable positions of movable units into a list
        for my_unit in my_movable_units:
            reachable = RangeController.get_reachable_cells_matrix(my_unit, map)
            for i in range(1, len(reachable)):
                for j in range(1, len(reachable[0])):
                    if reachable[i][j] and low_threshold <= risk_of_map[i][j] <= high_threshold:
                        move_actions.append(Action.create_move_only_action(my_unit, i, j))

        return move_actions

    @staticmethod
    def calc_risk_map(team_color, map):
        """
        Write the risk level to places within the range of enemy units' attacks.
        The risk level indicates the number of enemy units that can potentially attack that position.
        Higher values indicate higher risk (maximum is the number of attacks from 4 directions + the number of enemy self-propelled artillery).
        Note: The calculation does not consider the order of unit actions, so it may be slightly inaccurate at times.
        """
        dx = [+1, 0, -1, 0]
        dy = [0, -1, 0, +1]

        en_units = map.get_units_list(team_color, False, False, True)

        if_map = [[[False] * 5 for _ in range(map.get_y_size())] for _ in range(map.get_x_size()) for _ in range(len(en_units))]
        risk_of_map = [[0] * map.get_y_size() for _ in range(map.get_x_size())]

        i = 0

        for u in en_units:
            # For direct attack type units
            if u.get_spec().is_direct_attack_type():
                reachable_cells = RangeController.get_reachable_cells_matrix(u, map)

                for x in range(1, map.get_x_size() - 1):
                    for y in range(1, map.get_y_size() - 1):
                        # Check the positions above, below, left, and right of the corresponding cell
                        for r in range(4):
                            # If any of the above, below, left, or right positions are within the movable range, (x, y) is an attackable position
                            if reachable_cells[x + dx[r]][y + dy[r]]:
                                if_map[x][y][i][r] = True
            # For indirect attack units
            else:
                attackable_cells = RangeController.get_attackable_cells_matrix(u, map)

                for x in range(1, map.get_x_size() - 1):
                    for y in range(1, map.get_y_size() - 1):
                        # If (x, y) is an attackable position
                        if attackable_cells[x][y]:
                            if_map[x][y][i][4] = True
            i += 1

        for x in range(1, map.get_x_size() - 1):
            for y in range(1, map.get_y_size() - 1):
                # Initialization
                tor_f_unit = [False] * len(en_units)
                tor_f_direction = [False] * 4
                risk_of_unit = 0
                risk_of_line = 0
                risk_of_indirection = 0

                # Check the number of units that can attack the corresponding cell
                for k in range(len(en_units)):
                    tor_f_unit[k] = if_map[x][y][k][0] or if_map[x][y][k][1] or if_map[x][y][k][2] or if_map[x][y][k][3]
                    if tor_f_unit[k]:
                        risk_of_unit += 1
                    if if_map[x][y][k][4]:
                        risk_of_indirection += 1

                # Check the directions from which the corresponding cell can be attacked
                for r in range(4):
                    for k in range(len(en_units)):
                        tor_f_direction[r] |= if_map[x][y][k][r]
                    if tor_f_direction[r]:
                        risk_of_line += 1

                if risk_of_line > risk_of_unit:
                    risk_of_map[x][y] = risk_of_unit + risk_of_indirection
                else:
                    risk_of_map[x][y] = risk_of_line + risk_of_indirection

        return risk_of_map

    @staticmethod
    def calc_if_map(team_color, map):
        """
        Write the risk level to places within the range of enemy units' attacks in the risk level array.
        """
        dx = [+1, 0, -1, 0]
        dy = [0, -1, 0, +1]

        if_map = [[0] * map.get_y_size() for _ in range(map.get_x_size())]
        en_units = map.get_units_list(team_color, False, False, True)

        for u in en_units:
            # For direct attack type units
            if u.get_spec().is_direct_attack_type():
                reachable_cells = RangeController.get_reachable_cells_matrix(u, map)

                for x in range(1, map.get_x_size() - 1):
                    for y in range(1, map.get_y_size() - 1):
                        # Check the positions above, below, left, and right of the corresponding cell
                        for r in range(4):
                            # If any of the above, below, left, or right positions are within the movable range, (x, y) is an attackable position
                            if reachable_cells[x + dx[r]][y + dy[r]]:
                                if_map[x][y] += 1
            # For indirect attack units
            else:
                attackable_cells = RangeController.get_attackable_cells_matrix(u, map)

                for x in range(1, map.get_x_size() - 1):
                    for y in range(1, map.get_y_size() - 1):
                        # If (x, y) is an attackable position
                        if attackable_cells[x][y]:
                            if_map[x][y] += 1

        return if_map